### PR TITLE
Use fast rlp crate

### PR DESCRIPTION
**Pull request by vldm**
- [x] Wait for merge #32 
- [x]  Currently dependency is in local path
- [ ] We also can avoid copying hashes and nibble vecs, second one can help on short collections, while first probably can slightly improve perfomance on big branches.

Introduce next changes in perfomance
```
     Running benches/compare_backends.rs (target/release/deps/compare_backends-7340992dcd32978b)
Gnuplot not found, using plotters backend
bench get TrieCollection<Mem>
                        time:   [413.91 µs 416.98 µs 420.46 µs]
                        change: [-52.270% -51.735% -51.102%] (p = 0.00 < 0.05)
                        Performance has improved.

bench insert TrieCollection<Mem>
                        time:   [4.6423 ms 4.6650 ms 4.6871 ms]
                        change: [-51.300% -50.890% -50.493%] (p = 0.00 < 0.05)
                        Performance has improved.


```

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -83,7 +83,7 @@ impl Change {
         if node.inlinable() {
             MerkleValue::Full(Box::new(node.clone()))
         } else {
-            let subnode = rlp::encode(node).to_vec();
```
Comment (dj8yfo) [inline src/lib.rs:86]: Пирформанс аптиммзейшн спот

**Review by dj8yfo — CHANGES_REQUESTED** [review]

File: src/rlp/decode.rs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            "Not valid combination of payload_slize and list flag for MerkleValue, header",
+        ))
+    }
+}
+
+// Check if this rlp list consume all buffer, with given number of items.
+pub(crate) fn is_list_consume_rlp(mut buf: &[u8], num: usize) -> bool {
+    let buf = &mut buf;
+    for _i in 0..num {
+        let Ok(h) = fastrlp::Header::decode(buf) else {
```
Comment (dj8yfo) [inline src/rlp/decode.rs:97]: requires rust 1.65, please remove, so that it can be merged ASAP

File: src/rlp/decode.rs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            return false
+        };
+        if h.payload_length > buf.len() {
+            return false;
+        }
+        buf.advance(h.payload_length);
+    }
+    buf.is_empty()
+}
```
Comment (dj8yfo) [inline src/rlp/decode.rs:106]:
```
// this variant is a bit more precise.
// Err(conditions report if list is invalid altogether)
// Ok() - reports length of list, which the decision is based upon in outer match
pub(crate) fn count(mut buf: &[u8]) -> Result<usize, fastrlp::DecodeError> {
    let buf = &mut buf;
    let mut count: usize = 0;
    loop {
        if buf.is_empty() {
            break;
        }
        let Ok(h) = fastrlp::Header::decode(buf) else {
            return Err(fastrlp::DecodeError::Custom("header expected"));
        };
        if h.payload_length > buf.len() {
            return Err(fastrlp::DecodeError::InputTooShort);
        }
        buf.advance(h.payload_length);
        count += 1;

    }
    // are empty lists sick? (is empty list a valid case in rlp?)
    Ok(count)
}
```

File: src/rlp/decode.rs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+use bytes::Buf;
+pub use fastrlp::{Decodable, DecodeError};
+use primitive_types::H256;
+
+pub fn decode<'a, T: Decodable<'a>>(mut val: &'a [u8]) -> Result<T, DecodeError> {
+    Decodable::decode(&mut val)
+}
+
+impl<'a> MerkleNode<'a> {
+    pub fn fast_decode_with_header(
```
Comment (dj8yfo) [inline src/rlp/decode.rs:14]: rename to `decode_with_header` to further optimize reading

File: src/rlp/decode.rs (hunk 3 above)
Comment (dj8yfo) [inline src/rlp/decode.rs:106]:
```suggestion
// this variant is a bit more precise imho
// Err(conditions report if list is invalid altogether)
// Ok() - reports length of list, which the decision is based upon in outer match
pub(crate) fn count(mut buf: &[u8]) -> Result<usize, fastrlp::DecodeError> {
    let buf = &mut buf;
    let mut count: usize = 0;
    loop {
        if buf.is_empty() {
            break;
        }
        let Ok(h) = fastrlp::Header::decode(buf) else {
            return Err(fastrlp::DecodeError::Custom("header expected"));
        };
        if h.payload_length > buf.len() {
            return Err(fastrlp::DecodeError::InputTooShort);
        }
        buf.advance(h.payload_length);
        count += 1;

    }
    // are empty lists sick? (is empty list a valid case in rlp?)
    Ok(count)
}
```

File: src/rlp/mod.rs [hunk 5]
```diff
@@ -0,0 +1,252 @@
+pub mod decode;
+pub mod encode;
+#[cfg(feature = "old_rlp)")]
+pub mod old;
```
Comment (dj8yfo) [inline src/rlp/mod.rs:4]: somehow old cannot be made non-public
```
pub mod encode;
#[cfg(feature = "old_rlp)")]
pub mod old;

pub use decode::{Decodable, DecodeError as DecoderError};

pub use encode::Encodable;

use crate::merkle::nibble::{NibbleType, NibbleVec};

pub use encode::encode;

pub use decode::decode;

#[cfg(feature = "old_rlp)")]
mod testing {

    //!
    //! Implement marker traits for types that Decodable both from regular rlp, and from fastrlp.
    //! Usefull when need to test that both implementation has same output.
    //!

    trait Decodable<'a>: crate::rlp::Decodable<'a> + crate::rlp::old::Decodable<'a> {}
    impl<T> Decodable<'a> for T where T: crate::rlp::Decodable<'a> + crate::rlp::old::Decodable<'a> {}

    trait Encodable: crate::rlp::Encodable + crate::rlp::old::Encodable {}
    impl<T> Encodable for T where T: crate::rlp::Encodable + crate::rlp::old::Encodable {}
}

#[derive(Eq, PartialEq, Debug)]
pub struct NibblePair(pub NibbleVec, pub NibbleType);

#[cfg(test)]
mod tests {
    use crate::merkle::nibble::Nibble;
    use crate::merkle::{Branch, Extension, Leaf, MerkleNode, MerkleValue};
    use crate::{
        merkle::nibble::{self, NibbleType},
        rlp,
    };
    use hexutil::read_hex;
    use primitive_types::H256;
    use super::old;
```
```
  --> src/rlp/mod.rs:44:9
   |
44 |     use super::old;
   |         ^^^^^^^^^^ no `old` in `rlp`
```

File: Cargo.toml [hunk 6]
```diff
@@ -12,7 +12,9 @@ name = "triedb"
 
 [dependencies]
 sha3 = "0.9.1"
-rlp = "0.5.0"
+rlp_old = {version= "0.5.0", package="rlp", optional = true}
```
Comment (dj8yfo) [inline Cargo.toml:15]: Rename to old_rlp to remove indirection below?

**Comment by dj8yfo** [comment]
Please also consider removing `where Self: 'de` in fastrlp fork

**Review by dj8yfo — CHANGES_REQUESTED** [review]
After applying diff `old_rlp)` -> `old_rlp`
```
diff --git a/src/rlp/mod.rs b/src/rlp/mod.rs
index a18a1a5..cfe4e45 100644
--- a/src/rlp/mod.rs
+++ b/src/rlp/mod.rs
@@ -1,6 +1,6 @@
 pub mod decode;
 pub mod encode;
-#[cfg(feature = "old_rlp)")]
+#[cfg(feature = "old_rlp")]
 pub mod old;
 
 pub use decode::{Decodable, DecodeError as DecoderError};
@@ -13,7 +13,7 @@ pub use encode::encode;
 
 pub use decode::decode;
 
-#[cfg(feature = "old_rlp)")]
+#[cfg(feature = "old_rlp")]
 mod testing {
 
     //!
@@ -44,10 +44,10 @@ mod tests {
 
     macro_rules! check_roundtrip {
         ($v: expr => $type: ty) => {{
-            #[cfg(feature = "old_rlp)")]
+            #[cfg(feature = "old_rlp")]
             let old_rlp_raw: Vec<u8>;
             let rlp_raw;
-            #[cfg(feature = "old_rlp)")]
+            #[cfg(feature = "old_rlp")]
             {
                 old_rlp_raw = crate::rlp::old::encode(&$v);
                 dbg!(hexutil::to_hex(&old_rlp_raw));
@@ -61,7 +61,7 @@ mod tests {
                 assert_eq!(decoded_node, $v);
             }
 
-            #[cfg(feature = "old_rlp)")]
+            #[cfg(feature = "old_rlp")]
             {
                 assert_eq!(old_rlp_raw, rlp_raw);
             }
```
got a ton of compilation errors :
```
   Compiling triedb v0.5.0 (/home/jerryminion/Documents/code/study/rust/solana-slash-velas/triedb_raventid)
error[E0432]: unresolved import `super::NibblePair`
 --> src/rlp/old/decode.rs:7:5
  |
7 | use super::NibblePair;
  |     ^^^^^^^^^^^^^^^^^ no `NibblePair` in `rlp::old`

error[E0432]: unresolved import `super::NibblePair`
 --> src/rlp/old/encode.rs:6:5
  |
6 | use super::NibblePair;
  |     ^^^^^^^^^^^^^^^^^ no `NibblePair` in `rlp::old`

error[E0261]: use of undeclared lifetime name `'a`
  --> src/rlp/old/decode.rs:24:19
   |
24 | impl<T> Decodable<'a> for T
   |      -            ^^ undeclared lifetime
   |      |
   |      help: consider introducing lifetime `'a` here: `'a,`

error[E0261]: use of undeclared lifetime name `'a`
  --> src/rlp/mod.rs:25:23
   |
25 |     impl<T> Decodable<'a> for T where T: crate::rlp::Decodable<'a> + crate::rlp::old::Decodable<'a> {}
   |          -            ^^ undeclared lifetime
   |          |
   |          help: consider introducing lifetime `'a` here: `'a,`

error[E0261]: use of undeclared lifetime name `'a`
  --> src/rlp/mod.rs:25:64
   |
25 |     impl<T> Decodable<'a> for T where T: crate::rlp::Decodable<'a> + crate::rlp::old::Decodable<'a> {}
   |                                                                ^^ undeclared lifetime
   |
   = note: for more information on higher-ranked polymorphism, visit https://doc.rust-lang.org/nomicon/hrtb.html
help: consider making the bound lifetime-generic with a new `'a` lifetime
   |
25 |     impl<T> Decodable<'a> for T where T: for<'a> crate::rlp::Decodable<'a> + crate::rlp::old::Decodable<'a> {}
   |                                          +++++++
help: consider making the bound lifetime-generic with a new `'a` lifetime
   |
25 |     impl<T> Decodable<'a> for T where for<'a> T: crate::rlp::Decodable<'a> + crate::rlp::old::Decodable<'a> {}
   |                                       +++++++
help: consider introducing lifetime `'a` here
   |
25 |     impl<'a, T> Decodable<'a> for T where T: crate::rlp::Decodable<'a> + crate::rlp::old::Decodable<'a> {}
   |          +++

error[E0261]: use of undeclared lifetime name `'a`
  --> src/rlp/mod.rs:25:97
   |
25 |     impl<T> Decodable<'a> for T where T: crate::rlp::Decodable<'a> + crate::rlp::old::Decodable<'a> {}
   |                                                                                                 ^^ undeclared lifetime
   |
help: consider making the bound lifetime-generic with a new `'a` lifetime
   |
25 |     impl<T> Decodable<'a> for T where T: crate::rlp::Decodable<'a> + for<'a> crate::rlp::old::Decodable<'a> {}
   |                                                                      +++++++
help: consider making the bound lifetime-generic with a new `'a` lifetime
   |
25 |     impl<T> Decodable<'a> for T where for<'a> T: crate::rlp::Decodable<'a> + crate::rlp::old::Decodable<'a> {}
   |                                       +++++++
help: consider introducing lifetime `'a` here
   |
25 |     impl<'a, T> Decodable<'a> for T where T: crate::rlp::Decodable<'a> + crate::rlp::old::Decodable<'a> {}
   |          +++

warning: unused imports: `NibbleType`, `NibbleVec`
 --> src/rlp/old/mod.rs:7:29
  |
7 | use crate::merkle::nibble::{NibbleType, NibbleVec};
  |                             ^^^^^^^^^^  ^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error[E0592]: duplicate definitions with name `inlinable`
  --> src/rlp/decode.rs:54:5
   |
54 |     pub fn inlinable(&self) -> bool {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ duplicate definitions for `inlinable`
   |
  ::: src/rlp/old/encode.rs:14:5
   |
14 |     pub fn inlinable(&self) -> bool {
   |     ------------------------------- other definition for `inlinable`

Some errors have detailed explanations: E0261, E0432, E0592.
For more information about an error, try `rustc --explain E0261`.
warning: `triedb` (lib) generated 1 warning
error: could not compile `triedb` due to 7 previous errors; 1 warning emitted
```
Please either remove `old_rlp` or make it compile.

File: src/rlp/mod.rs (hunk 5 above)
Comment (dj8yfo) [inline src/rlp/mod.rs:4]: please make it non-`pub`

**Comment by Tirka** [comment]
Bump crate to version `0.6.0` after merge?

**Review by dj8yfo — APPROVED** [review]

**Pull request merged**